### PR TITLE
Support BUILD_PATH_PREFIX_MAP in ocamldebug

### LINKED ISSUE #12083: Modify ocamldebug to handle Dune workspace mapping
In order to produce reproducible builds (independent of the location of the build), starting from Dune 3.0, Dune has mapped references to the workspace directory to “/workspace_root”. The source/load path directories contained in debug information produced by ocamlc were affected by this change, for users who build their projects with Dune. The result is that ocamldebug sometimes is unable to find the program source code.

Starting in Dune 3.7, a dune-project stanza was added to suppress the workspace mapping. 
```
(map_workspace_root <bool>)
```

This enables users to get programs that ocamldebug can handle, at the expense of not having reproducible builds.

Another workaround is for the ocamldebug user to add `-I <dir>` options showing where the sources are.

Note that if the project uses bin and lib directories, as suggested by `dune init project`, then the source file references in the debug information will be things like "bin/main.ml" or "lib/mod.ml". Currently, the emacs "ocamldebug" command that is used
to start ocamldebug, changes to the directory of the bytecode file, e.g. _build/default/bin. From that directory, it will not
be able to find the sources. If, before trying to do anything that causes the program to load, the user uses "cd" to move up to the directory containing "_build", then the relative paths in the debug events will work ok. That is yet another workaround for users.

This feature request proposes an ocamldebug change that is relatively simple and would work most of the time. Since it is Dune that is causing the problem, it will not occur unless the user is using Dune to build their project. In that case, the build artifacts are contained in a "_build" directory (at least by default). ocamldebug can look at the location of the binary being debugged (or of additional binaries loaded dynamically). If they are contained in a "_build" directory, then the occurrences of "/workspace_root" can be mapped to the directory containing the _build directory. That should restore the directories to what they were without the mapping. 

I have a version of ocamldebug that has been so modified and plan to submit a PR soon.

### PR DESCRIPTION
**Pull request by richardlford**
In order to produce reproducible builds (independent of the location
of the build), starting from Dune 3.0, Dune has mapped references to
the workspace directory to “/workspace_root”.

This change enables ocamldebug to recover the locations of the source
code. The ocamldebug user must set the BUILD_PATH_PREFIX_MAP
environment variable to effect the mapping.

If we only want to map "/workspace_root" to a single directory, say
mydir, we would do:

export BUILD_PATH_PREFIX_MAP="mydir=/workspace_root"

If we want to map to multiple directories, we must separate
them, and the trailing part of the path, with ":', but
in a BUILD_PATH_PREFIX_PATH these are encoded as "%.".
So to map to mydir1 and mydir2, the setting would be:

export BUILD_PATH_PREFIX_MAP="mydir1%.mydir2%.=/workspace_root"

Then if ocamldebug has a path like

"/workspace_root/stuff",

The initial mapping takes this to

"mydir1:mydir2:/stuff"

Then ocamldebug will convert this to the list

["mydir1/stuff" "mydir2/stuff"]
Fixes #12083

### CODE REVIEW

**Comment by dra27** [comment]
There's discussion still in the original issue, so this doesn't necessarily want amending yet, but a few comments:
- `_build` is also used by ocamlbuild, so from the compiler's perspective, it's not a strong enough heuristic
- The tests would need updating to reference `str` properly, but I don't think we want that dependency - I think it's better to split the path on the separator and manipulate the resulting `string list` instead
- I'm (perhaps unsurprisingly!) sceptical as to whether the Windows handling of the path is correct 🙂 (`/` vs `\` etc)

**Comment by richardlford** [comment]
@dra27, I've picked a simple alternative to using regular expressions, so do not need Str. I also changed to use either "/" or "\" as file separator. I also rebased.

**Comment by richardlford** [comment]
As mentioned in https://github.com/ocaml/ocaml/issues/12083#issuecomment-1469777516, I've decided to use `BUILD_PATH_PREFIX_PATH` in ocamldebug. I will be revising my changes accordingly and pushing the revised version when ready.

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: I've push a commit that reverts my prior implementation and instead uses BUILD_PATH_PREFIX_PATH. The user (or build environment) is responsible for setting this appropriately.
The ocaml implementation supports mapping to multiple directories, as is necessary if some artifacts
are in the local build and some in the install directory.

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: I've pushed a commit to:
1. Fix whitespace problem
2. Add Changes entry
3. Add a section in the manual describing BUILD_PATH_PREFIX_PATH usage.

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: I've push a commit to update dependencies (using "make alldepend").

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: I've pushed a commit to fix the manual build as well as add more details to the manual.

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: Is there anyone special that would review the manual changes?

**Comment by richardlford** [comment]
@dra27, @gasche, @dbuenzli: I think I'm done for today. I think the implementation is complete. The things I see remaining are:

- Diagnose why a test is failing. It does not fail for me locally. Could it be Windows related?
- Write some tests for the test suite. I need to figure out how to do that. I have tested the funcitonality manually and it all seems to be working fine.

I'll resume these efforts tomorrow. If you have any review comments that would be helpful. For example, do you like the idea of a debugger variable to control BUILD_PATH_PREFIX_MAP from within the debugger? I thought it was a nice addition. Any comments on the manual or man page changes?

**Comment by shindere** [comment]
Isn't it `BUILD_PATH_PREFIX_MAP` that you mean, rather than `BUILD_PATH_PREFIX_PATH`?

**Comment by shindere** [comment]
Richard L Ford (2023/03/15 05:56 -0700):
> @dra27, @gasche, @dbuenzli: I've pushed a commit to:
> 1. Fix whitespace problem
> 2. Add Changes entry
> 3. Add a section in the manual describing BUILD_PATH_PREFIX_PATH
> usage.

Would you mind rewriting the history to keep only what's relevant,
please?

**Comment by shindere** [comment]
Richard L Ford (2023/03/15 09:36 -0700):
> @dra27, @gasche, @dbuenzli: Is there anyone special that would review
> the manual changes?

@Octachron maybe? But it's no use to repeatedly mention people IMO.

**Comment by shindere** [comment]
Richard L Ford (2023/03/15 15:21 -0700):
> @dra27, @gasche, @dbuenzli: I think I'm done for today. I think the implementation is complete. The things I see remaining are:
> 
> - Diagnose why a test is failing. It does not fail for me locally.
> - Could it be Windows related?

Will have a look into it, today, tomorrow or at the  beginning  of last
week.

> - Write some tests for the test suite. I need to figure out how to do that. I have tested the funcitonality manually and it all seems to be working fine.
> 
I can help with that one. Feel free to contact me on Slack or by email
and we can even organize an audio meeting.

> I'll resume these efforts tomorrow. If you have any review comments
> that would be helpful. For example, do you like the idea of a debugger
> variable to control BUILD_PATH_PREFIX_MAP from within the debugger? I
> thought it was a nice addition. Any comments on the manual or man page
> changes?

Did you have a look to how other debuggers deal withthis and howthey
give access to the feature? It may be good to take inspiration there.

I think the `ocamldebug` guru is @damiendoligez, who may also have an
opinion.

**Comment by richardlford** [comment]
I squashed and force-pushed.

**Comment by richardlford** [comment]
> sn't it `BUILD_PATH_PREFIX_MAP` that you mean, rather than `BUILD_PATH_PREFIX_PATH`?

You are right. I hope I have fixed all the places where I had that.

**Comment by richardlford** [comment]
> But it's no use to repeatedly mention people IMO.

Sorry, I'm not sure what the etiquette is. If someone has contributed to a PR conversation, do they automatically get notified if there are any changes, without being mentioned again?

**Comment by shindere** [comment]
Richard L Ford (2023/03/16 01:45 -0700):
> > But it's no use to repeatedly mention people IMO.
>
> Sorry, I'm not sure what the etiquette is. If someone has contributed
> to a PR conversation, do they automatically get notified if there are
> any changes, without being mentioned again?

People can configure whether they get notified for pushes or not.

Independently of that, the way people are notified of comments is up to
them. I think most of the core developers of OCaml are `watching` the
repository, so they see everything, be it trhoug web notificaitons or
through email. Then, as soon has they have been mentionned once on a
PR/issue/discussion, or as soon as they participate, the e-mails they
receive are sent in Cc as opposed to those sent for threads one is not
participating to, which are sent in Bcc.

This happens  until threads are closed or people unsubscribe from them.

**Comment by richardlford** [comment]
I pushed another commit that completes the work for the PR. 
- I added 4 tests
-  I enhanced ocamltest with a new "expand" action, that copies (expanding ocamltest variables) from a source to destination file. I needed this because my mappings needed to use ocamltest variable contents in files.
- I fixed a couple of places in the compiler where it was missing sanitizing directories destined for the debug information.

Since this includes a bit of new code, and especially since it includes compiler changes, I'm expecting it will need further review. In the meantime, if that one test is still failing on Windows, I'll look into it. My new tests were based on that test, so I'm quite familiar with it.

File: parsing/location.ml [hunk 1]
```diff
@@ -174,6 +174,23 @@ let absolute_path s = (* This function could go into Filename *)
...
+  (* Now simplify . and .. components *)
+  let rec aux s =
+    let base = basename s in
+    let dir = dirname s in
+    if dir = s then dir
+    else if base = current_dir_name then aux dir
+    else if base = parent_dir_name then dirname (aux dir)
+    else concat (aux dir) base
+  in
+  aux s
```
Comment (gasche) [inline parsing/location.ml]: I don't like the amount of code duplication with the function `absolute_path` above, and I am not fond of the name either -- `_always` suggests that we always do something, but what?

If we need to have this feature of deciding locally to rewrite only relative paths or also absolute paths, I would propose the following interface:

```
val absolute_path : rewrite:[ `Absolute | `All ] -> string -> string
```

**Comment by richardlford** [comment]
> - Write some tests for the test suite. I need to figure out how to do that. I have tested the funcitonality manually and it all seems to be working fine.
> I can help with that one. Feel free to contact me on Slack or by email and we can even organize an audio meeting.

@shindere, I found your helpful ocamltest tutorial. With that and reading the ocamltest sources I was able to write my tests. As mentioned above, I added a new ocamltest action, so you may be interested in reviewing it (or suggest a way I could have done without it).

**Review by gasche — COMMENTED** [review]
Generally the work here is impressive, you are touching many different parts of the tools with apparent ease and nice documentation. I have a few concerns with the current proposal however:

1. There are too many things mixed together that need to be looked at by different people. Your second commit (1) adds an entirely new mode of use of BUILD_PATH_PREFIX_MAP in the Location API and (2) uses it in various places of the compiler and (3) implements a new ocamltest feature (4) uses them in new tests. This is way too much stuff in a single commit and in fact arguably it would be better in separate PRs. My proposal would be to:
- agree on the support we want for BUILD_PATH_PREFIX_MAP and merge that, without tests for now
- send the `expand` ocamltest action as a follow-up PR, along with ocamldebug tests using it servince as a justification for the change

2. I am not sure about the design using `:` to cram multiple potential output paths for a single input path. I think that `:` in the format is meant to allow users to specify paths themselves containing the `:` character, for example on Windows. We discussed in the past having an option to "invert" the semantics of the BUILD_PATH_PREFIX_MAP file, and I would note that this suffices to specify one-to-many rewrite relations: if you write a map that maps both `foo` and `bar` to `/workspace_root`, then playing the "reverse" mapping will naturally return both `foo/a` and `bar/a` for the input `/workspace_root/a`. (But maybe there are other approaches.)

3. I don't like your approach regarding `Location.absolute_path_always`. When we introduced BUILD_PATH_PREFIX_MAP we discussed whether to rewrite only relative paths or all kind of paths. I think I made the choice to rewrite only relative paths to make the change less invasive / more controlled. Maybe that was the wrong choice, as @dbuenzli apparently belivies, and we could revisit it and discuss applying the rewriting on absolute paths as well. However, your current suggestion feels like a hack to avoid having this discussion: you add both modes (rewrite only relative paths, rewrite all paths), and we start randomly using one or the other inside the codebase depending on what tests tell us that we should probably do. This increases the total complexity of the system, I would like to avoid this extra complexity.

**Comment by dbuenzli** [comment]
> Maybe that was the wrong choice, as @dbuenzli apparently belivies, and we could revisit it and discuss applying the rewriting on absolute paths as well. 

It's not about a wrong choice. It's just that relative paths are not that much of a problem for reproducible builds if you build from a consistent location and only use relative paths (which I don't consider to be a good idea from a build usability point of view). 

E.g. if you have your `cwd` in say `_build` and only use relative paths from there, everyone will have the same paths for a given build.

It's absolute paths that are problematic and that's precisely the ones that OCaml doesn't support…

**Comment by richardlford** [comment]
@gasche: I agree. Here are some comments on your review.

> 1. There are too many things ...

I agree. I was just trying to get something complete but am happy to
break it up.

> My proposal would be to:
> 
> * agree on the support we want for BUILD_PATH_PREFIX_MAP and merge that, without tests for now
> * send the `expand` ocamltest action as a follow-up PR, along with ocamldebug tests using it servince as a justification for the change

Agreed.

> 2. I am not sure about the design using `:` to cram multiple
  potential output paths for a single input path. I think that `:` in
  the format is meant to allow users to specify paths themselves
  containing the `:` character, for example on Windows. We discussed
  in the past having an option to "invert" the semantics of the
  BUILD_PATH_PREFIX_MAP file, and I would note that this suffices to
  specify one-to-many rewrite relations: if you write a map that maps
  both `foo` and `bar` to `/workspace_root`, then playing the
  "reverse" mapping will naturally return both `foo/a` and `bar/a` for
  the input `/workspace_root/a`. (But maybe there are other
  approaches.)

Let me think. Let's assume first of all, as we previously discussed,
that the build-time layout does not contradict the install-time
layout. The possibility of contradiction occurs because source locations
in the debug information come from two sources. There are the
directories to look in, but also each debug event has a relative file
location that may include a directory part. And currently, the way
Dune works at least, unless the source file is in the top level, there
will be a directory part of such a relative path.

Currently, at least, Dune maps the root of the build tree
(e.g. srcroot/_build/default) to /workspace_root.

Let consider two possible source layouts. 

Layout I:

All libraries are in directories directly below the top, but they may
be in different packages.  Then the compilation of file `liba/moda.ml`
(i.e. `liba/moda.cmo`) currently has these debug directories:

- `/workspace_root`
- `/workspace_root/liba`

and the file part of debug events would be `liba/moda.ml`. If these
are part of `pkga`, the source code, and `liba.cma` will be installed at

`opamroot/lib/pkga/liba/moda.ml`

And similarly for pkgb, pkgc, etc. Let's suppose that our current
workspace involves packages `pkga`, `pkgb`, and `pkgc`, where we are
using `pkga` from the installed location, `pkgb` is installed, but we
want a local copy, and `pkgc` is not installed and is a work in
progress. An execuable in `pkgc` will have modules from all three
packages.

In this case, no rewrite of the plain `/workspace_root` directory will
let us find the sources, since it misses the package part of the
installation path. However,

- if we rewrite `/workspace_root/liba` to `opamroot/lib/pkga/`, then
  that will find the sources that we want from the library,

- if we rewrite `/workspace_root/liba` to `mysourcedir/`,
  then that will find the sources in our local build directory.

We would need one such mapping for each library (multiple per package
if a package had multiple libraries), choosing to map either to the
opam library, or to the local source directory on a case-by-case
basis. That could be a lot of work!

In layout I, there is no advantage to being able to map to multiple
directories.

Layout II:

In this layout, below the top level of the workspace there is a
directory for each package that will be installed, with the same name
as its installation name. The libraries contained in a package will be
below it. For example,

The compilaton of file `pkga/liba/moda.ml` currently has these debug
directories:

- `/workspace_root`
- `/workspace_root/pkga/liba`

and the file part of debug events would be `pkga/liba/moda.ml`. The
installation location will be

`opamroot/lib/pkga/liba/moda.ml`

And similarly for pkgb, pkgc, etc. As before, let's suppose that our
current workspace involves packages `pkga`, `pkgb`, and `pkgc`, where
we are using `pkga` from the installed location, `pkgb` is installed,
but we want a local copy, and `pkgc` is not installed and is a work in
progress. An execuable in `pkgc` will have modules from all three
packages.

Unlike Layout I, in this case, rewriting the plain `/workspace_root`
directory will let us find the sources.

- if we rewrite `/workspace_root` to `opamroot/lib`, then that will
  find the sources that we want from the library,

- if we rewrite `/workspace_root` to `mysourcedir`, then that will
  find the sources in our local build directory.

But we will want to give priority to our source directory, then to our
local build directory (e.g. _build/default), and then to the opam
library. We can do this by using the feature of mapping a single
abstract directory to multiple directories. That is the justification
I have for the feature of mapping to multiple directories. It makes
managing the mappings simple.

But it only works with Layout II. Otherwise, you have to resort to
very complex and cumbersome mappings. So that is also why I advocate
for Layout II.

However, as you mentioned above (and I had forgotten), Windows
uses ":", so we would have to pick some other delimiter to separate 
the parts in the mapping.

> 3. I don't like your approach regarding
  `Location.absolute_path_always`. When we introduced
  BUILD_PATH_PREFIX_MAP we discussed whether to rewrite only relative
  paths or all kind of paths. I think I made the choice to rewrite
  only relative paths to make the change less invasive / more
  controlled. Maybe that was the wrong choice, as @dbuenzli apparently
  belivies, and we could revisit it and discuss applying the rewriting
  on absolute paths as well. However, your current suggestion feels
  like a hack to avoid having this discussion: you add both modes
  (rewrite only relative paths, rewrite all paths), and we start
  randomly using one or the other inside the codebase depending on
  what tests tell us that we should probably do. This increases the
  total complexity of the system, I would like to avoid this extra
  complexity.

The issue is whether you want any paths in your output that are
specific to your local environment. Currently, we map the local user
workspace, but do not map the location of the opam directory, so the
absolute paths to your opam directory will leak into your outputs.

If Layout II were used, then both your local build root as well as the
local `opam/switch/lib` could be mapped to `/workspace_root` and the
reverse mapping I suggest above would work.

Of course, the compiler can be willing to map absolute paths other
than the `cwd` of the build, but if the build system does not setup a
mapping for them, nothing will happen.

In the compiler changes I was making, I wanted to be careful not do
map to abstract paths when the compiler wanted the absolute path to
actually use, e.g. to read from. That is why I wrote the variant that
I did.

**Comment by mobileink** [comment]
A few tidbits from the bazel guy.

There is no `_build` directory under Bazel. Build outputs generally go into a dir structure that mirrors the source structure, but relative to directories that Bazel generates.

Similarly for inputs.  Every input must be registered with Bazel, which does a bunch of symlinking so that when an action runs its inputs are relative to a bazel-defined dir. Which means incidentally that if you fail to register an input file, it will be ignored even if it exists where you expect it.  Example just to illustrate the kind of path it creates:

```
/private/var/tmp/_bazel_gar/289dc3938ecab7c846ccaf2627af4b2f/execroot/__main__/external/mibl/ -> /Users/gar/obazl/mibl
```
And that path can vary even for files in the same directory, since you can dynamically change the build configuration, and Bazel will put each build config in its own space.

I guess the lesson here is: never use absolute paths unless you really really, no really, know what you are doing.

I'm not really sure what this implies for this proposal since I don't know much about  the debugger. Maybe BUILD_PATH_PREFIX_MAP would suffice.

What I do know is that Dune is not the One True Build system and should not be treated as such (monoculture bad!).

Something else I know: this is not the only place where tooling puts src paths into outputs in a bad way.  Some ppxes and some tests in the ocaml repo do this, resulting in non-replicable builds. The problem is that the tests/ppxes as run by dune put the filename into the output, I think (recalling) because dune does a chdir before running stuff.  But with bazel the outputs contain a full path (I don't remember if it is absolute or relative).  So anything involving diffs breaks.  I've been able to work around this for some ppxes by passing `-loc-filename`.  In any case, it strikes me as a kind of general tool design issue for which a standard recommendation or policy might be advisable.

**Comment by mobileink** [comment]
And here's an impertinent question. Also embarrassing since I should know the answer but I don't: why are we even discussing running the debugger against installed libs?  I've noticed the profligate used of `-g` but I've eliminated it in the bazel rules because it makes no sense to me and I haven't gotten around to investigating further.  I did check the dune build log for a few packages, and it looks like `-g` is used pretty much everywhere - not because the dune file specified it but evidently dune thinks its a good idea. Why? If you want `-g`, run a debug build, full stop.  It doesn't belong in a standard build, much less an optimized build, which is what I would expect for e.g. opam packages.  What am I missing?

**Comment by richardlford** [comment]
> Also embarrassing since I should know the answer but I don't: why are we even discussing running the debugger against installed libs?

Good question. I can perhaps think of a couple of reasons:

1. For production work, people are probably using native code, so the question is mute, and they are not losing any efficiency because they are not using the bytecode version.
2. I, and perhaps others, find it educational to be able to step from their application into library code, to understand what it is doing. On the other hand, it is sometimes annoying to try to get past the system code initialization to your "own code". If the libraries had no debug information then your first stop would be in your own code. Personally, I'm always curious, so I like having sources and being able to step into installed libraries. One of the great advantages of open-source is being able to learn the mysteries of other people's code. And that sometimes makes up less-than-perfect documentation.

Regarding getting to your "own code", the problem I've encountered is that ocamldebug's facility for setting a breakpoint at a function rarely works, because it can only be executed in a context where the target function is in-scope, i.e. needed for some purpose. And that rarely happens. Fortunately, I have in my pipeline a fix for that. There is enough information in the debug information to know before execution starts where all of the functions are. So I soon plan to submit a PR for ocamldebug that will let you set a function breakpoint whenever you want, and particularly before starting the program.

**Comment by richardlford** [comment]
Regarding multiple target paths, I've had some enlightenment.

First of all, it should have been obvious that ':' would not work on
Windows. I've pushed a change to use `Misc.split_path_contents` instead,
i.e. ";" on Windows and ':' on other systems.

Also, I was incorrect in saying that multiple targets would not
benefit "Layout I". Let me explain.

The path derived from the directories in the debug information is
used for two things: finding sources, but also finding binaries like
the ".cma" files. So we should consider both. Let's consider the two
scenarios again.

For brevity, let's abbreviate:

- `srcroot` = The user's development directory
- `binroot` = The build system's directory for build artifacts that also
  mirrors the sources. For dune it would be somethings like
  `srcroot/_build/context`.
- `libroot` = The directory below which packages are installed. For opam
  this would be `~/.opam/<your-switch>/lib`.

Layout I:

All libraries are in directories directly below `srcroot`, but they may
be in different packages.  Then the compilation of file
`srcroot/liba/moda.ml` (i.e. `binroot/liba/moda.cmo`) currently has these
debug directories:

- `/workspace_root`
- `/workspace_root/liba`

and the file part of debug events would be `liba/moda.ml`. If these
are part of `pkga`, the source code will be installed at

`libroot/pkga/liba/moda.ml`

and `liba.cma` will be installed at

`libroot/pkga/liba/liba.cma`

In this case, no _single_ rewrite of the plain `/workspace_root`
directory will let us find the sources, since it misses the package
part of the installation path. However, we can map `/workspace_root`
to these multiple targets in decreasing order:

- `srcroot`, to pick up primary sources rather than those in binroot.
- `binroot`, to pick up binary artifacts locally in preference to
  those in `libroot`.
- Having a mapping to just `libroot` is not helpful, as it misses the
  packages part, but
- For each used library package, `pkgx`, map to
  - `libroot/pkgx`

The above mappings will find all the appropriate sources and binaries
without the need to map on the specific `/workspace_root/liba`.

Layout II:

In this layout, below the `srcroot` there is a directory for each
package that will be installed, with the same name as its installation
name. The libraries contained in a package will be below it. For
example,

The compilaton of file `srcroot/pkga/liba/moda.ml` currently has these
debug directories:

- `/workspace_root`
- `/workspace_root/pkga/liba`

and the file part of debug events would be `pkga/liba/moda.ml`. The
installation location will be

`libroot/pkga/liba/moda.ml` and

`libroot/pkga/liba/liba.cma`

Now the following multiple rewrites of `/workspace_root` will work:

- `srcroot`, to pick up primary sources rather than those in binroot.
- `binroot`, to pick up binary artifacts locally in preference to
  those in libroot.
- `libroot` will find all the sources and binaries needed
- There no need to have additional mappings for individual installed
  packages. 

In conclusion, multiple mappings are useful. They are simplest with
Layout II, but also useful with Layout I.

**Comment by richardlford** [comment]
@shindere, I need some help with my tests on Windows. To set the value of BUILD_PATH_PREFIX_MAP I was currently using some lines like 

```
set BUILD_PATH_PREFIX_MAP="/stdlib_root=${ocamlsrcdir}/stdlib"
```

The problem is that on Windows, `${ocamlsrcdir}` may expand to something with colons. The colons need to be encoded as "%.". So far I do not see an easy way to do that. Currently, the compiler is failing when it tries to use the invalid BUILD_PATH_PREFIX_MAP. 

One idea is to write a script that takes its argument and a target file, and then does this encoding and writes the result to the target file. But how would I transfer the value from the file back into a variable?

Or maybe enhance the current "script" action, so if an "output_variable" parameter was set to the name of a variable, then the output from stdout would be captured and stored in that variable?

Or I've thought it might be handy if ocamltest had a concept of built-in functions. So maybe something like:

```
set BUILD_PATH_PREFIX_MAP="/stdlib_root=${encode ${ocamlsrcdir}}/stdlib"
```
where "encode" would be a built-in function to do the encoding. The variable registry could be enhanced to allow marking a variable as a function and attaching an ocaml function to it. 

Or alternatively something like

```
myvar = `script args`
```
using shell-like back-tick notation to run a script and return the result.

I would also like to be able to run the CI environment locally so I can catch these problems before pushing changes. 
Is there a Windows VM (I'm using VmWare) available like the OCaml CI uses? 

Also, what is the Windows environment like. I see that tests are using the `sh` shell. Is this POSIX shell, or like bash? Would the "tr" utility be available?

Thanks for your help.

**Comment by richardlford** [comment]
I have made more changes, but also squash everything into one. As @gasche suggested, this will be split up into multiple PRs (at least 2), but for now, I'm just trying to get a clean test run (on Windows). When that is clean, I will work on splitting it up.

The changes since last time:

1. I decided the root cause of the Windows failures was that the BUILD_PATH_PREFIX_MAP values were not being encoded as needed, and the ":"s on Windows were causing problems. 
2. I added a facility for builtin functions to ocamltest and defined two functions and tests for them:
- bppm_encode, to do the encoding (needed for my tests), and
- bppm_decode, for decoding (tested but not currently used)
3. I added the `dump_expanded` action to ocamltest in order to see the expanded values of the variables. The plain `dump` action shows the unexpanded values.
4. I modified my BUILD_PATH_PREFIX_MAP tests to use bppm_encode.

**Comment by richardlford** [comment]
I pushed another commit.

I decided the problem I was addressing for the debugger
was more general and that the BUILT_PATH_PREFIX_MAP
spec should be generalized to handle it.

I captured the current spec on my github account, then made 
a branch with my proposed changes.
See https://github.com/richardlford/build-path-prefix-map-spec/pull/1.

1. Modified BUILT_PATH_PREFIX_MAP handling in the compiler accordingly
2. Used the new facilities in ocamldebug
3. Updated one test for the new spec.
4. Modify encoding and decoding functions in ocamltest.

**Comment by richardlford** [comment]
Thanks to all who have given feedback so far. I think the functionality is now correct, so I will proceed to break it into the following pieces:

1. The compiler and ocamldebug changes to support an enhanced BUILT_PATH_PREFIX_MAP specification
2. Enhancements to ocamltest
3. Tests of no. 1 that motivated no. 2.

I'm going to leave this PR in its current state until those are completed so that reviewers can see all three if desired.
